File: HoT-frontend/src/services/patients.ts
import { API_BASE_URL } from '../config';
import { Patient, VitalSigns } from '../types';

export const PatientService = {
    async getAllPatients(): Promise<Patient[]> {
        try {
            const response = await fetch(`${API_BASE_URL}/api/patients`);
            if (!response.ok) {
                throw new Error('Failed to fetch patients');
            }
            return await response.json();
        } catch (error) {
            console.error('Error fetching patients:', error);
            throw error;
        }
    },

    async getPatientByID(patientID: string): Promise<Patient> {
        try {
            const response = await fetch(`${API_BASE_URL}/api/patients/${patientID}`);
            if (!response.ok) {
                throw new Error('Failed to fetch patient');
            }
            return await response.json();
        } catch (error) {
            console.error('Error fetching patient:', error);
            throw error;
        }
    },

    async getPatientVitals(patientID: string): Promise<VitalSigns> {
        try {
            const response = await fetch(`${API_BASE_URL}/api/patients/${patientID}/vitals`);
            if (!response.ok) {
                throw new Error('Failed to fetch patient vitals');
            }
            return await response.json();
        } catch (error) {
            console.error('Error fetching patient vitals:', error);
            throw error;
        }
    }
};

File: HoT-backend/internal/api/handlers/patients.go
package handlers

import (
    "encoding/json"
    "net/http"
    "github.com/gorilla/mux"
    "github.com/sshreyx1/HealthofThings/HoT-backend/internal/services"
)

type PatientHandler struct {
    patientService *services.PatientService
}

func NewPatientHandler(patientService *services.PatientService) *PatientHandler {
    return &PatientHandler{
        patientService: patientService,
    }
}

func (h *PatientHandler) GetAllPatients(w http.ResponseWriter, r *http.Request) {
    patients, err := h.patientService.GetAllPatients()
    if err != nil {
        http.Error(w, err.Error(), http.StatusInternalServerError)
        return
    }

    w.Header().Set("Content-Type", "application/json")
    json.NewEncoder(w).Encode(patients)
}

func (h *PatientHandler) GetPatientByID(w http.ResponseWriter, r *http.Request) {
    vars := mux.Vars(r)
    patientID := vars["id"]

    patient, err := h.patientService.GetPatientByID(patientID)
    if err != nil {
        http.Error(w, err.Error(), http.StatusNotFound)
        return
    }

    w.Header().Set("Content-Type", "application/json")
    json.NewEncoder(w).Encode(patient)
}

func (h *PatientHandler) GetPatientVitals(w http.ResponseWriter, r *http.Request) {
    vars := mux.Vars(r)
    patientID := vars["id"]

    vitals, err := h.patientService.GetPatientVitals(patientID)
    if err != nil {
        http.Error(w, err.Error(), http.StatusNotFound)
        return
    }

    w.Header().Set("Content-Type", "application/json")
    json.NewEncoder(w).Encode(vitals)
}